src/databricks/labs/ucx/hive_metastore/mapping.py

```diff
@@ -103,14 +125,90 @@ def skip_table(self, backend: StatementExecutionBackend, schema: str, table: str
+        for value in self._backend.fetch(f"DESCRIBE SCHEMA EXTENDED {database}"):
+            describe[value["database_description_item"]] = value["database_description_value"]
+        if self.UCX_SKIP_PROPERTY in TablesCrawler.parse_database_props(describe.get("Properties", "").lower()):
+            logger.info(f"Database {database} is marked to be skipped")
+            return None
+        return database
+
+    def _get_tables_in_scope(self, rules: list[Rule], databases_in_scope: set[str], crawled_tables: list[Table]):
+        crawled_tables_keys = {crawled_table.key: crawled_table for crawled_table in crawled_tables}
+        upgraded_tables = self._get_upgraded_tables()
+        tables_to_check = []
+        for rule in rules:
+            if rule.as_hms_table_key not in crawled_tables_keys:
+                logger.info(f"Table {rule.as_hms_table_key} in the mapping doesn't show up in assessment")
+                continue
+            if rule.as_hms_table_key in upgraded_tables:
+                logger.info(f"Table {rule.as_hms_table_key} was migrated to {rule.as_uc_table_key} and will be skipped")
+                continue
+            if rule.src_schema not in databases_in_scope:
+                logger.info(f"Table {rule.as_hms_table_key} is in a database that was marked to be skipped")
+                continue
+            tables_to_check.append(TableToMigrate(crawled_tables_keys[rule.as_hms_table_key], rule))
+        tasks = []
+        for table_to_check in tables_to_check:
+            tasks.append(partial(self._get_table_in_scope_task, table_to_check))
+        return Threads.strict("checking all database properties", tasks)
+
+    def _get_table_in_scope_task(self, table_to_migrate: TableToMigrate) -> TableToMigrate | None:
+        table = table_to_migrate.src
+        rule = table_to_migrate.rule
+
+        result = self._backend.fetch(f"SHOW TBLPROPERTIES `{table.database}`.`{table.name}`")
+        for value in result:
+            if value["key"] == self.UCX_SKIP_PROPERTY:
+                logger.info(f"{table.key} is marked to be skipped")
+                return None
+            if value["key"] == "upgraded_to":
+                logger.info(f"{table.key} is set as upgraded to {value['value']}")
+                if self._is_target_exists(value["value"]):
+                    return None
+                logger.info(f"The upgrade target for {table.key} is missing. Unsetting the upgrade_to property")
+                self._backend.execute(table.sql_unset_upgraded_to())
+        if self._is_target_exists(rule.as_uc_table_key):
+            logger.info(f"The intended target for {table.key}, {rule.as_uc_table_key}, already exists.")
+            return None
+        return table_to_migrate
+
+    def _is_target_exists(self, target_key: str):
+        # Attempts to get the target table info from UC returns True if it exists.
+        try:
+            self._ws.tables.get(target_key)
+            return True
```

Comment:
```suggestion
        if self._exists_in_uc(rule):
            logger.info(f"The intended target for {table.key}, {rule.as_uc_table_key}, already exists.")
            return None
        return table_to_migrate

    def _exists_in_uc(self, rule: Rule):
        # Attempts to get the target table info from UC returns True if it exists.
        try:
            table_info = self._ws.tables.get(rule.as_uc_table_key)
            upgraded_from = table_info.properties.get('upgraded_from', None)
            if upgraded_from and upgraded_from != rule.as_hms_table_key:
              raise ResourceConflict(f"Expected to be migrated from {rule.as_hms_table_key}, but got {upgraded_from}. You can skip this error via command: databricks labs ucx skip --schema {rule.src_schema} --table {rule.src_table}")
            return True
```

We don't have transactional integrity, so that's why we have to do this extra validation here.

p.s.: `_is_target_exists` - is not a correct English

Reply:
We need to check both ends. I'll make it more elegant.

Reply:
We have to account for multiple conditions:
We have upgraded_from on a UC table (we should skip)
We have upgraded_to on the HMS table and it's pointing nowhere (we should revert and migrate)
We have upgraded_to on the HMS table and it's pointing to an existing UC table (we should skip)
We have an existing UC table at the target (we should skip)

Reply:
I don't understand why the upgraded_from on the target is material. If we have an object at the target, we don't upgrade.

Reply:
I left it as is (renamed the method). Let's discuss.

Reply:
Fixed

Reply:
Fixed